siad/daemon.go

```diff
@@ -28,137 +25,49 @@ type DaemonConfig struct {
-		return
-	}
-
-	d.state = consensus.CreateGenesisState()
-	d.gateway, err = gateway.New(config.RPCAddr, d.state, filepath.Join(config.SiaDir, "gateway"))
+// newDaemon initializes modules using the config parameters and uses them to
+// create an api.Server.
+func newDaemon(cfg DaemonConfig) (d *daemon, err error) {
+	state := consensus.CreateGenesisState()
+	gateway, err := gateway.New(cfg.RPCAddr, state, filepath.Join(cfg.SiaDir, "gateway"))
 	if err != nil {
 		return
 	}
-	d.tpool, err = transactionpool.New(d.state, d.gateway)
+	tpool, err := transactionpool.New(state, gateway)
 	if err != nil {
 		return
 	}
-	d.wallet, err = wallet.New(d.state, d.tpool, filepath.Join(config.SiaDir, "wallet"))
+	wallet, err := wallet.New(state, tpool, filepath.Join(cfg.SiaDir, "wallet"))
 	if err != nil {
 		return
 	}
-	d.miner, err = miner.New(d.state, d.gateway, d.tpool, d.wallet)
+	miner, err := miner.New(state, gateway, tpool, wallet)
 	if err != nil {
 		return
 	}
-	d.host, err = host.New(d.state, d.tpool, d.wallet, filepath.Join(config.SiaDir, "host"))
+	host, err := host.New(state, tpool, wallet, filepath.Join(cfg.SiaDir, "host"))
 	if err != nil {
 		return
 	}
-	d.hostdb, err = hostdb.New(d.state, d.gateway)
+	hostdb, err := hostdb.New(state, gateway)
 	if err != nil {
 		return
 	}
-	d.renter, err = renter.New(d.state, d.gateway, d.hostdb, d.wallet, filepath.Join(config.SiaDir, "renter"))
+	renter, err := renter.New(state, gateway, hostdb, wallet, filepath.Join(cfg.SiaDir, "renter"))
 	if err != nil {
 		return
 	}
 
-	// Subscribe to the miner and wallet to keep all parts of the daemon
-	// synchronized.
-	d.minerUpdates = d.miner.MinerSubscribe()
-	d.walletUpdates = d.wallet.WalletSubscribe()
-
-	// Register RPCs for each module
-	d.gateway.RegisterRPC("AcceptBlock", d.acceptBlock)
-	d.gateway.RegisterRPC("AcceptTransaction", d.acceptTransaction)
-	d.gateway.RegisterRPC("HostSettings", d.host.Settings)
-	d.gateway.RegisterRPC("NegotiateContract", d.host.NegotiateContract)
-	d.gateway.RegisterRPC("RetrieveFile", d.host.RetrieveFile)
-
-	d.initAPI(config.APIAddr)
-
-	return
-}
-
-func (d *daemon) updateWait() {
-	<-d.minerUpdates
-	<-d.walletUpdates
-}
-
-func createSubdirs(rootDir string) error {
-	subdirs := []string{"gateway", "wallet", "host", "renter"}
-	for _, subdir := range subdirs {
-		err := os.MkdirAll(filepath.Join(rootDir, subdir), 0777)
-		if err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
-// TODO: move this to the state module?
-func (d *daemon) acceptBlock(conn modules.NetConn) error {
-	var b consensus.Block
-	err := conn.ReadObject(&b, consensus.BlockSizeLimit)
-	if err != nil {
-		return err
-	}
-
-	err = d.state.AcceptBlock(b)
-	if err == consensus.ErrOrphan {
-		go d.gateway.Synchronize(conn.Addr())
-		return err
-	} else if err != nil {
-		return err
+	// bootstrap to the network
+	// TODO: probably a better way of doing this.
+	if !config.Siacore.NoBootstrap {
+		go gateway.Bootstrap(modules.BootstrapPeers[0])
 	}
```

Comment:
should the bootstrapping be handled by the gateway? I feel like it's not really the daemon's job, it's more of the gateway's job.

The daemon is really just a tiny shell that strings everything together.